homeassistant/components/api/__init__.py

```diff
@@ -367,20 +366,27 @@ async def post(self, request, domain, service):
 
         Returns a list of changed states.
         """
-        hass = request.app["hass"]
+        hass: ha.HomeAssistant = request.app["hass"]
         body = await request.text()
         try:
             data = json.loads(body) if body else None
         except ValueError:
             return self.json_message("Data should be valid JSON.", HTTP_BAD_REQUEST)
 
-        with AsyncTrackStates(hass) as changed_states:
-            try:
-                await hass.services.async_call(
-                    domain, service, data, blocking=True, context=self.context(request)
-                )
-            except (vol.Invalid, ServiceNotFound) as ex:
-                raise HTTPBadRequest() from ex
+        context = self.context(request)
+
+        try:
+            await hass.services.async_call(
+                domain, service, data, blocking=True, context=context
+            )
+        except (vol.Invalid, ServiceNotFound) as ex:
+            raise HTTPBadRequest() from ex
+
+        changed_states = []
+
+        for state in hass.states.async_all():
+            if state.context is context:
+                changed_states.append(state)
```

Comment:
Isn't this prone to race conditions. If a state change triggers another (and doesn't inherit context). The context might have changed before we get here?

Reply:
If context is not inherited and it's happening while the service is being called, it's not caused by the service call and so shouldn't be included.

Reply:
Okay true. But what if the state changes again due to other reasons after it has changed due to this call. Say a updatecoordinator triggers update of entities just after this service call triggered an update. SInce service call can be long running, the first state changes it caused, may get overwritten by dataupdatecoordinators triggering their running.

Reply:
That's a general issue with the context implementation though, isn't it?
We don't really know if a state change is the result of a certain service call, the entity will simply attribute any state changes within a given time window to the context.

Reply:
I much rather prefer to totally remove this logic, it's not used besides here, and this code is not used by HA 🤷

Reply:
The problem with the old approach is that we hugely over-report, especially for long running tasks.

Reply:
So this comment is blocking this PR from moving forward. Since it's not used and an artifact from 2014 when the frontend still relied on this 🤷

Reply:
Feels like removal would be a good choice. Old logic overreport, new logic underreport.

If this i legacy. I don't really mind either way. So don't consider it blocking.